pydantic_ai_slim/pydantic_ai/_parts_manager.py

```diff
@@ -67,6 +69,52 @@ def get_parts(self) -> list[ModelResponsePart]:
         """
         return [p for p in self._parts if not isinstance(p, ToolCallPartDelta)]
 
+    def __stop_tracking_vendor_id(self, vendor_part_id: VendorId | None) -> None:
+        """Stop tracking a vendor_part_id (no-op if None or not tracked)."""
+        if vendor_part_id is not None:  # pragma: no branch
+            self._vendor_id_to_part_index.pop(vendor_part_id, None)
+
+    def _append_part(self, part: ManagedPart, vendor_part_id: VendorId | None = None) -> int:
+        """Append a part, optionally track vendor_part_id, return new index."""
+        new_index = len(self._parts)
+        self._parts.append(part)
+        if vendor_part_id is not None:
+            self._vendor_id_to_part_index[vendor_part_id] = new_index
+        return new_index
+
+    def _latest_part_isinstance_of(self, *part_types: type[PartT]) -> tuple[PartT, int] | None:
+        """Get the latest part and its index if it's an instance of the given type(s)."""
+        if self._parts:
+            part_index = len(self._parts) - 1
+            latest_part = self._parts[part_index]
+            if isinstance(latest_part, part_types):
+                return latest_part, part_index
+        return None
+
+    def _handle_embedded_thinking_start(
+        self, vendor_part_id: VendorId, provider_details: dict[str, Any] | None
+    ) -> Iterator[ModelResponseStreamEvent]:
+        """Handle <think> tag - create new ThinkingPart."""
+        self.__stop_tracking_vendor_id(vendor_part_id)
```

Comment:
Do we need this? Won't `_append_part` overwrite the key in `_vendor_id_to_part_index` anyway?

Reply:
I'd rather leave it in, in the future some logic may come in between that and appending (`if sth: don't append`) and this line will save someone an hour of debugging :D